### PR TITLE
TextField with custom label

### PR DESCRIPTION
**Pull request by alfetopito**
Depends on #2 

# Update Aug 13

More stories, updated storybook version, hiding error messages, highlighting `warn` and `error` based on component flags:

screenshot_2020-08-13_14-04-32

---

First base component to implement design defined on https://projects.invisionapp.com/share/VXXWLEE7534#/screens/427462292_GP_CMM_-_V1_1-3_v5_DeployStrategy-START-_1

Built on top of Safe React Components `TextField` component. Hides the inner label and allows a custom label on top of it.

Example storybook:
screenshot_2020-08-11_09-20-18

### CODE REVIEW

File: src/components/Inputs/TextFieldWithCustomLabel/index.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+// To match parent's width https://github.com/gnosis/safe-react-components/blob/development/src/inputs/TextField/index.tsx#L24
+const WIDTH = 400;
+
+const Wrapper = styled.div<{ width?: string | number }>`
+  width: ${({ width }) => pxOrCustomCssUnits(width || WIDTH)};
+
+  // Adjust input padding which is normally pushed down to accommodate default label 
+  .MuiFilledInput-input {
+    padding 11px 12px 9px;
```
Comment (alfetopito) [inline src/components/Inputs/TextFieldWithCustomLabel/index.tsx:16]: Not happy with this. Doesn't feel right.
Also, doesn't align nicely when adding `endAdornment`.

How can I do this better, the proper way?

File: src/components/Inputs/TextFieldWithCustomLabel/index.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    ${({ width }) => (width ? `width: ${pxOrCustomCssUnits(width)}` : "")}
+  }
+`;
+
+export interface Props
+  extends Omit<React.ComponentProps<typeof TextField>, "label"> {
+  customLabel: React.ReactElement;
+}
+
+export const TextFieldWithCustomLabel: React.FC<Props> = (props) => {
```
Comment (mikheevm) [inline src/components/Inputs/TextFieldWithCustomLabel/index.tsx:29]: React.FC type has some downfalls, for example, it adds an implicit type for children

More info: https://github.com/facebook/create-react-app/pull/8177

File: src/components/Inputs/TextFieldWithCustomLabel/TextFieldWithCustomLabel.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    <form noValidate autoComplete="off" onSubmit={onSubmit}>
+      <TextFieldWithCustomLabel
+        {...textFieldData}
+        customLabel={undefined}
+        value={value}
+        onChange={(e) => setValue(e.target.value)}
+      />
+    </form>
+  );
+};
```
Comment (mikheevm) [inline src/components/Inputs/TextFieldWithCustomLabel/TextFieldWithCustomLabel.stories.tsx:84]: An input without label is considered bad practice (we've talked about that briefly in slack), I think it's better to remove this example and enforce customLabel prop to be defined

File: src/components/Inputs/TextFieldWithCustomLabel/index.tsx (hunk 1 above)
Comment (mikheevm) [reply src/components/Inputs/TextFieldWithCustomLabel/index.tsx:16]: Seems like the way to go: https://material-ui.com/components/text-fields/#CustomizedInputs.tsx (Bootstrap input)

**Comment by mikheevm** [comment]
Good job, I think it would be great to have this component in safe-react-components

**Comment by alfetopito** [comment]
@mikheevm I think there's a problem with this implementation.
I did not account for error messages, so it shows up all mixed up with the input:
screenshot_2020-08-13_08-42-23


In our use case, we'll not show the error message in the component. Instead, we'll highlight it, and show the messages somewhere else, like this:
screenshot_2020-08-13_08-38-26

But this makes the component less generic.
If I were still to push this to `safe-react-components`, what would you expect the behaviour of the error message to be?

**Comment by mikheevm** [comment]
@alfetopito would it be possible to include the error message inside the input (like it currently is)?

**Comment by alfetopito** [comment]
> @alfetopito would it be possible to include the error message inside the input (like it currently is)?

And it that case, you'd expect the input to shift back down like it was before when using the label?

Without error
screenshot_2020-08-14_08-38-52

With error (plus the label, not in this image but imagine it is :) )
screenshot_2020-08-14_08-39-30

**Comment by mikheevm** [comment]
@alfetopito  Oh, I understand now. I think it would be good if we can avoid the jump. Not sure how to do it the best way. Maybe we can just move the input down a little bit (not sure how it will look tho)

**Comment by alfetopito** [comment]
> @alfetopito Oh, I understand now. I think it would be good if we can avoid the jump. Not sure how to do it the best way. Maybe we can just move the input down a little bit (not sure how it will look tho)

If we move the input down, it will no longer be centered when there's no error msg to display

**Comment by alfetopito** [comment]
I'll merge as is but we can continue the discussion regarding porting this back to `safe-react-components`

**Pull request merged**